### DISCUSSION TITLE
Calling kfuncs from bpftrace

### DISCUSSION THREAD

**Discussion by jfernandez** [bpftrace/bpftrace#3167] [General]
Is it possible to call a kfunc from bpftrace? In particular `bpf_task_from_pid`.

I'm working on a runqlat program and I need to filter out some tasks based on the info the the task struct. The `sched_new` and `sched_switch` hooks only provide the pids.

**Comment by jfernandez** [comment]
I learned that bpftrace support raw tracepoints, which provide the task structs as params.
```
rawtracepoint:sched_wakeup,
rawtracepoint:sched_wakeup_new,
{
    $task = (struct task_struct *)arg0;
    printf("Wakeup %d %s\n", $task->pid, $task->comm);
}
```

I'm still interested in finding out if kfunc calls are supported.

**Comment by ajor** [comment]
Calling arbitrary kfuncs isn't something that's supported in bpftrace right now, but I can see it being useful functionality to have. I'm not quite sure how we'd expose this in the language yet.

If it's still something you want, would you mind opening a "feature request" issue for it?

**Reply by jfernandez** [reply]
https://github.com/bpftrace/bpftrace/issues/3169

**Comment by danobi** [comment]
The machinery for calling kfuncs is blocked on https://github.com/bpftrace/bpftrace/issues/2334

Whether or not we want to allow arbitrary calls is interesting but a separate discussion. It sounds awfully like adding support for https://github.com/bpftrace/bpftrace/issues/2976 (in a good way). Definitely something we can consider.

Historically, we allow-list helpers via implementing bpftrace helpers. But since kfuncs are unstable, it could make sense to offload compat to the user. It's a bit of a tradeoff IMO

**Reply by jordalgo** [reply]
Do we have the BTF for these kfuncs and does that make it easier to implement this (after https://github.com/bpftrace/bpftrace/issues/2334 is resolved of course) ?

**Reply by danobi** [reply]
Yes, we do. Both:

* https://lore.kernel.org/bpf/cover.1706491398.git.dxu@dxuuu.xyz/
* https://lore.kernel.org/bpf/cover.1714430735.git.dxu@dxuuu.xyz/

have been merged. And at some point in the future we will be able to feature probe kfuncs.
